### DISCUSSION TITLE
<MilvusException: (code=2, message=Fail connecting to server on localhost:19530. Timeout)>

### DISCUSSION THREAD

**Discussion by jeff62802217** [milvus-io/milvus#27830] [Q&A and General discussion]
When I try to build up a connection to milvus on a new server,  I fail:

`
from pymilvus import connections

connections.connect(alias=('default) host='localhost', port='19530')
`

I've checked that the milvus service was running by the container(docker ps -a | grep milvus) :
`
e1b0ba65325c   milvusdb/milvus:v2.3.0                                                    "/tini -- milvus run_"   43 minutes ago   Up 43 minutes (healthy)     0.0.0.0:9091->9091/tcp, :::9091->9091/tcp, 0.0.0.0:19530->19530/tcp, :::19530->19530/tcp                                                milvus-standalone
117cf9f21c5a   minio/minio:RELEASE.2023-03-20T20-16-18Z                                  "/usr/bin/docker-ent_"   43 minutes ago   Up 43 minutes (healthy)     0.0.0.0:9000-9001->9000-9001/tcp, :::9000-9001->9000-9001/tcp                                                                           milvus-minio
9602406a7123   quay.io/coreos/etcd:v3.5.5                                                "etcd -advertise-cli_"   43 minutes ago   Up 43 minutes (healthy)     2379-2380/tcp                                                                                                                           milvus-etcd
`

The build up docker file is downloaded from "wget https://github.com/milvus-io/milvus/releases/download/v2.3.0/milvus-standalone-docker-compose.yml -O docker-compose.yml":

`
version: '3.5'

services:
  etcd:
    container_name: milvus-etcd
    image: quay.io/coreos/etcd:v3.5.5
    environment:
      - ETCD_AUTO_COMPACTION_MODE=revision
      - ETCD_AUTO_COMPACTION_RETENTION=1000
      - ETCD_QUOTA_BACKEND_BYTES=4294967296
      - ETCD_SNAPSHOT_COUNT=50000
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/etcd:/etcd
    command: etcd -advertise-client-urls=http://127.0.0.1:2379 -listen-client-urls http://0.0.0.0:2379 --data-dir /etcd
    healthcheck:
      test: ["CMD", "etcdctl", "endpoint", "health"]
      interval: 30s
      timeout: 20s
      retries: 3

  minio:
    container_name: milvus-minio
    image: minio/minio:RELEASE.2023-03-20T20-16-18Z
    environment:
      MINIO_ACCESS_KEY: minioadmin
      MINIO_SECRET_KEY: minioadmin
    ports:
      - "9001:9001"
      - "9000:9000"
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/minio:/minio_data
    command: minio server /minio_data --console-address ":9001"
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
      interval: 30s
      timeout: 20s
      retries: 3

  standalone:
    container_name: milvus-standalone
    image: milvusdb/milvus:v2.3.0
    command: ["milvus", "run", "standalone"]
    environment:
      ETCD_ENDPOINTS: etcd:2379
      MINIO_ADDRESS: minio:9000
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/milvus:/var/lib/milvus
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9091/healthz"]
      interval: 30s
      start_period: 90s
      timeout: 20s
      retries: 3
    ports:
      - "19530:19530"
      - "9091:9091"
    depends_on:
      - "etcd"
      - "minio"

networks:
  default:
    name: milvus
`

Can anyone help figure out why the connection to milvus failed?

**Comment by yhmo** [comment]
1. Use this script to test connection
```
from pymilvus import connections
connections.connect(host='localhost', port='19530')
print(utility.get_server_version())
```

2. Extract the log of the milvus container, attach the last 1000 lines log here. 
`docker logs e1b0ba65325c > 1.log`

**Reply by jeff62802217** [reply]
2023/10/20 07:10:55 maxprocs: Leaving GOMAXPROCS=112: CPU quota undefined

    __  _________ _   ____  ______    
   /  |/  /  _/ /| | / / / / / __/    
  / /|_/ // // /_| |/ / /_/ /\ \    
 /_/  /_/___/____/___/\____/___/     

Welcome to use Milvus!
Version:   v2.3.0-dev
Built:     Wed Aug 23 14:42:13 UTC 2023
GitCommit: 8d3398c1
GoVersion: go version go1.20.7 linux/amd64

open pid file: /run/milvus/standalone.pid
lock pid file: /run/milvus/standalone.pid
WARNING: Logging before InitGoogleLogging() is written to STDERR
I20231020 07:10:55.677878    86 knowhere_config.cc:93] [KNOWHERE][SetBlasThreshold][milvus] Set faiss::distance_compute_blas_threshold to 16384
I20231020 07:10:55.677961    86 knowhere_config.cc:104] [KNOWHERE][SetEarlyStopThreshold][milvus] Set faiss::early_stop_threshold to 0
I20231020 07:10:55.677989    86 knowhere_config.cc:52] [KNOWHERE][ShowVersion][milvus] Knowhere Version: v2.2.0


The log's output is as the above, so it seems the milvus version is not compatible?

I typed "conda list | grep milvus" and the output is :

milvus                    2.3.0                    pypi_0    pypi
pymilvus                  2.3.0                    pypi_0    pypi

**Reply by jeff62802217** [reply]
The output of connection script is like this:

>>> connections.connect(host='localhost', port='19530')
Traceback (most recent call last):
  File "/data/anaconda3/envs/siran_br1toN/lib/python3.8/site-packages/pymilvus/client/grpc_handler.py", line 131, in _wait_for_channel_ready
    grpc.channel_ready_future(self._channel).result(timeout=timeout)
  File "/data/anaconda3/envs/siran_br1toN/lib/python3.8/site-packages/grpc/_utilities.py", line 151, in result
    self._block(timeout)
  File "/data/anaconda3/envs/siran_br1toN/lib/python3.8/site-packages/grpc/_utilities.py", line 97, in _block
    raise grpc.FutureTimeoutError()
grpc.FutureTimeoutError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/anaconda3/envs/siran_br1toN/lib/python3.8/site-packages/pymilvus/orm/connections.py", line 355, in connect
    connect_milvus(**kwargs, user=user, password=password, token=token, db_name=db_name)
  File "/data/anaconda3/envs/siran_br1toN/lib/python3.8/site-packages/pymilvus/orm/connections.py", line 302, in connect_milvus
    gh._wait_for_channel_ready(timeout=timeout)
  File "/data/anaconda3/envs/siran_br1toN/lib/python3.8/site-packages/pymilvus/client/grpc_handler.py", line 134, in _wait_for_channel_ready
    raise MilvusException(
pymilvus.exceptions.MilvusException: <MilvusException: (code=2, message=Fail connecting to server on localhost:19530. Timeout)>

**Reply by yhmo** [reply]
The log you show is incomplete. The milvus is written by GO and C++, the go version is the version we compile the source code. Your sdk is python sdk, it has nothing to do with go.

The error "Fail connecting to server on localhost:19530. Timeout" is client-side error, it means the client cannot connect to the milvus server. But the client doesn't know why the server cannot connect.

Only the full log can know why the milvus server failed.

**Make sure the client is not blocked by your machine's firewall.**

**Comment by RamanHrynkevich** [comment]
Hello.

I have a similar problem.
I am trying to start milvus and fastapi server with docker-compose.

But when I try to connect to milvus via fastapi I get the same error.

But when trying to connect from outside the container, the connection goes fine.

docker-compose :

`version: '3.5'

services:
  etcd:
    container_name: milvus-etcd
    image: quay.io/coreos/etcd:v3.5.5
    networks:
          - milvus
    environment:
      - ETCD_AUTO_COMPACTION_MODE=revision
      - ETCD_AUTO_COMPACTION_RETENTION=1000
      - ETCD_QUOTA_BACKEND_BYTES=4294967296
      - ETCD_SNAPSHOT_COUNT=50000
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/etcd:/etcd
    command: etcd -advertise-client-urls=http://127.0.0.1:2379 -listen-client-urls http://0.0.0.0:2379 --data-dir /etcd
    healthcheck:
      test: ["CMD", "etcdctl", "endpoint", "health"]
      interval: 30s
      timeout: 20s
      retries: 3

  minio:
    container_name: milvus-minio
    image: minio/minio:RELEASE.2023-03-20T20-16-18Z
    networks:
      - milvus
    environment:
      MINIO_ACCESS_KEY: minioadmin
      MINIO_SECRET_KEY: minioadmin
    ports:
      - "9001:9001"
      - "9000:9000"
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/minio:/minio_data
    command: minio server /minio_data --console-address ":9001"
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
      interval: 30s
      timeout: 20s
      retries: 3

  standalone:
    container_name: milvus-standalone
    image: milvusdb/milvus:v2.3.0
    networks:
      - milvus
    command: ["milvus", "run", "standalone"]
    environment:
      ETCD_ENDPOINTS: etcd:2379
      MINIO_ADDRESS: minio:9000
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/milvus:/var/lib/milvus
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9091/healthz"]
      interval: 30s
      start_period: 90s
      timeout: 20s
      retries: 3
    ports:
      - "19530:19530"
      - "9091:9091"
    depends_on:
      - "etcd"
      - "minio"

  fastapi:
    build: ./bots_api
    networks:
      - milvus
    environment:
      MILVUS_HOST: '0.0.0.0'
      MILVUS_PORT: '19530'
    depends_on:
      - standalone
    ports:
      - "5000:5000"

networks:
    milvus:

`

**Reply by yhmo** [reply]
Try MILVUS_HOST: 'standalone'

**Reply by RamanHrynkevich** [reply]
Thank you for your reply.
Unfortunately, nothing has changed.

**Reply by yhmo** [reply]
Hmmm... Is the "fastapi" container based on a Linux OS image? If so, check the connection by these steps:
1. use `docker exec -it [container id] `to enter the "fastapi" container
2. use commands `apt update`, `apt install iputils-ping`, `apt install telnet` to install tools
3. use command `ping standalone` to ping the standalone container. if no problem, it prints the ip of this container
4. use command `telnet xx.xx.xx.xx 19530` to test the milvus port. If no problem, it says "Connected xx.xx.xx.xx 19530"

**Reply by yhmo** [reply]
Another test:
1. add an ubuntu container into the docker-compose.yaml
```
  ubuntu:
    container_name: ubuntu-os
    image: ubuntu:latest
    tty: true
```

2. use `docker exec -it [container id]` to enter the "ubuntu" container
3. `apt-update`
4. `apt install python3 python3-pip vim`
5. `python3 -m pip install pymilvus`
6. create a python file xxx.py, edit it by the vim
7. paste the code into the python file
```
from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
)

connections.connect(host='standalone', port='19530')
print(utility.get_server_version())
print(utility.list_collections())
```
8. `python3 ./xxx.py`
9. the xxx.py prints these lines:
```
v2.3.0
[]
```

If all these steps are ok, the root cause is in your container "fastapi".

**Reply by RamanHrynkevich** [reply]
Can't get any further than point 8.
I get an error: 

[Errno 2] No such file or directory

**Reply by yhmo** [reply]
The step 6~8, use these command to create/edit a file:
`touch xxx.py` ------- create an empty file
`vim ./xxx.py` ---------- enter vim
Press "I" button ------ enter insert edit mode
Paste the code into the file
Press "ESC" button ---- exit edit mode
Press "Shift"+":" button, type "wq" and press "Enter" button ----- exit vim
`python3 ./xxx.py` -------- python runs the xxx.py script

**Reply by RamanHrynkevich** [reply]
It worked, the answer is this:


v2.3.2-dev
['raa_question']

**Reply by RamanHrynkevich** [reply]
raa_question is a collection I created earlier from the outside.

**Reply by RamanHrynkevich** [reply]
Like you said, the problem was on the FastAPI side and turned out to be quite simple....

It was necessary to specify host for connection not lockalhost, but the name of the container.

This option gave an error:
connections.connect(alias="default", host='localhost', port='19530')

But this one worked:
connections.connect(alias="default", host='standalone', port='19530')

**Reply by yhmo** [reply]
Right, because each container has a different ip address. Connect 'localhost' inside the FastAPI container actually is connecting itself. The 'standalone' container has different ip.
